Add experimental volumes and volumeMounts properties to jobs

radu-matei opened the pull request:
depends on https://github.com/brigadecore/brigadier/pull/22

This PR enables experimental support for adding volumes and volume mounts to jobs. 

Note that for the simple use case of sharing a volume across all jobs, `storage.enabled` should still be used.

A pre-built image for the worker: `radumatei/brigade-worker:volume-mounts-check`.

cc @krancour - this is a very Kubernetes-specific addition - how would this impact non-Kubernetes targets?

Use cases: https://github.com/brigadecore/brigade/pull/955#issuecomment-518002970 and  [Mounting volumes for ETL jobs](https://hackmd.io/@-IwbEKEMQBWBBEpIBone5g/Bk5zvEJEV)

Example:

```
    const kind = new Job("kind", "radumatei/golang-kind:1.11-0.4")
    kind.privileged = true;
    kind.volumes = [
        {
            name: "modules",
            hostPath: {
                path: "/lib/modules",
                type: "Directory"
            }
        },
        {
            name: "cgroup",
            hostPath: {
                path: "/sys/fs/cgroup",
                type: "Directory"
            }
        },
        {
            name: "docker-graph-storage",
            emptyDir: {}
        }
    ];

    kind.volumeMounts = [
        {
            name: "modules",
            mountPath: "/lib/modules",
            readOnly: true
        },
        {
            name: "cgroup",
            mountPath: "/sys/fs/cgroup"
        },
        {
            name: "docker-graph-storage",
            mountPath: "/var/lib/docker"
        }
    ];
```

radu-matei commented on brigade-worker/src/k8s.ts:
```diff
@@ -442,6 +442,16 @@ export class JobRunner implements jobs.JobRunner {
+      if (cfg.volume.hostPath != undefined && !project.allowHostMounts) {
+        throw new Error(`allowHostMounts is false in this project, not mounting path ${cfg.volume.hostPath.path}`)
+      }
+      this.runner.spec.volumes.push(cfg.volume);
```
I think this should check for unique volumes?

What is the default Kubernetes behaviour if:
- a volume name is present multiple times
- a PVC is present multiple times
- a host path mount is present multiple times

mumoshu replied:
I have never got to do it myself. But I see some use with subpath, as mentioned in the issue.

Just curious, but does whether to support mounting the same volume multiple times affect our design/implementation? 🤔

mumoshu replied:
Ah I got it! You've paired a volume and its corresponding mount in each volumeConfig.

I'd rather suggest exposing `volumes` and `volumeMounts` without any abstraction first, so that we can provide a standard library feature like this `volumeConfig` on top of that, or users can build their own.

IMHO for the standard library, something that solves 80% of the use-cases will be enough, that justifies dropping support for mounting the same volume multiple times.

mumoshu replied:
Reading https://github.com/radu-matei/brigadier/commit/94f56c2446c67e817e69730ea7cf6e9f2cdd3db0 - If you agree, implementation-wise, I think adding `volumeMounts` and `volumes` that are allowed/used only when `volumeConfig` doesn't exist would suffice.

radu-matei replied:
That totally makes sense - I initially thought maybe we could simplify and boil things down to a single property, but yeah, I you're right, we should keep in line with the Kubernetes abstraction.

Updating.

radu-matei replied:
I actually don't think we should keep both `volume` + `volumeMounts` and `volumeConfig`, as it just invites confusion - what do you think?

mumoshu replied:
Btw I like the idea of pairing volumes and mounts.

I even think we can happily merge the volume and the mount as in many cases each volume is mounted only once:

```
job.volumeConfig = [
    {
        name: "modules",
        mountPath: "/lib/modules",
        readOnly: true,
        hostPath: {
            path: "/lib/modules",
            type: "Directory"
        }
    },

    {
        name: "cgroup",
        mountPath: "/sys/fs/cgroup",
        hostPath: {
            path: "/sys/fs/cgroup",
            type: "Directory"
        }
    },
    {
        name: "docker-graph-storage",
        mountPath: "/var/lib/docker",
        emptyDir: {}
    }
]
```

And I believe `name`s can be usually omitted because they can be generated from mount-path(which are guaranteed to not collide:

```
job.volumeConfig = [
    {
//        name: "lib-modules",
        mountPath: "/lib/modules",
        readOnly: true,
        hostPath: {
            path: "/lib/modules",
            type: "Directory"
        }
    },

    {
//        name: "sys-fs-cgroup",
        mountPath: "/sys/fs/cgroup",
        hostPath: {
            path: "/sys/fs/cgroup",
            type: "Directory"
        }
    },
    {
//        name: "var-lib-docker",
        mountPath: "/var/lib/docker",
        emptyDir: {}
    }
]
```

mumoshu replied:
> I actually don't think we should keep both volume + volumeMounts and volumeConfig, as it just invites confusion - what do you think?

So I'd vote for starting with only `volumes` + `volumeMounts` if I need to choose one.

// But I do believe something like `volumeConfig` has a lot of potential, and worth it, even if it needs an extensive documentation to not confuse users.

radu-matei replied:
Yeah, I initially thought that something like `volumeConfig` could be an abstraction over how volumes and volume mounts work in Kubernetes and simplify setting them up for a job.

But I agree, let's start with plain `volumes` and `volumeMounts`, then decide whether to add a higher level abstraction, how you suggested.

Thanks a lot, this has been really helpful!

radu-matei replied:
Brigadier change: https://github.com/brigadecore/brigadier/pull/22/commits/8c53a1c91bf165c69ff8205712800f120223b82b

radu-matei replied:
Updated - could you please give this another look?
Thanks!

radu-matei commented on brigade-worker/package.json:
```diff
@@ -38,7 +38,7 @@
   },
   "dependencies": {
-    "@brigadecore/brigadier": "^0.5.0",
+    "@brigadecore/brigadier": "radu-matei/brigadier#job-volume-config-out",
```
After https://github.com/brigadecore/brigadier/pull/22 is merged and a new version is published, update this.

radu-matei commented on brigade-worker/src/k8s.ts:
```diff
@@ -12,6 +12,7 @@ import * as fs from "fs";
 import * as path from "path";
 import * as request from "request";
 import * as byline_1 from "byline";
+import { V1Volume } from "@kubernetes/client-node";
```
Remove this.

vdice commented on brigade-worker/src/k8s.ts:
```diff
@@ -442,6 +443,23 @@ export class JobRunner implements jobs.JobRunner {
+    }
+
+    // If the job defines volume mounts, add them to every container's spec.
+    for (let m of job.volumeMounts) {
```
Since `volumeMounts` and `volumes` appear to be un-coupled, should we add the same check here to be sure `project.allowHostMounts` is true?

And/or, in general, should we somehow check for proper API usage?  (If, say, configured with a volumeMount setup but no corresponding volume defined, etc. -- or is that allowed?)  I suppose as long as we can show the user the k8s library error that should result on mis-configuration, we'd be good.

radu-matei replied:
Since you need to declare both volume and volume mount, I think checking is unnecessary - it would end up just taking the volume mount name and iterating through the volumes to check if it's mounting a host path - which already do.

As for checking the existence of the volume you reference in the volume mount, that is a good idea, I'll add that.

radu-matei replied:
Added a check for the referenced volume, together with a test.

radu-matei replied:
Pushed `radumatei/brigade-worker:volume-mounts-check` if you need to try it on a project.

mumoshu replied:
I love the extra validation with `volumeExists` because that provides us a faster feedback loop 👍

vdice replied:
Perfect, thanks Radu!

radu-matei commented:
Merged https://github.com/brigadecore/brigadier/pull/22, rebasing, updating, and adding a doc.

krancour commented:
> cc @krancour - this is a very Kubernetes-specific addition - how would this impact non-Kubernetes targets?

Do we imagine being able to one day orchestrate job containers using something other than Kubernetes? It wasn't on my radar, but even if the answer were yes, I think this could be accommodated with other orchestrators.

radu-matei commented:
Rebased and added a simple doc.

@krancour - not really, or at least not the JS worker anyway. Just trying to make sure we're not overlooking something that might be affected by this - thanks!

radu-matei commented:
Thanks a lot for the help, everyone!
